docs/proposals/gateway-settings.md

```diff
@@ -0,0 +1,175 @@
+
+## Goals
+
+- Define the Gateway settings.
+- Define an API for the Gateway settings.
+
+## Non-Goals
+
+- Provide implementation details for implementing the Gateway settings.
+
+## Introduction
+
+### Gateway Settings
+
+Gateway Settings are NGINX directives or configuration attached at the GatewayClass level that should be solely controlled by the Cluster Operator and should not be changed by the Application Developers. All Gateways attached to this GatewayClass will inherit these settings.
+
+These settings apply to the `main`, `http`, and/or `stream` contexts of the NGINX config.
+
+To begin, the Gateway Settings config will include the following NGINX directives (focusing on OpenTelemetry tracing):
+
+- [`otel_exporter`](https://nginx.org/en/docs/ngx_otel_module.html#otel_exporter)
+- [`otel_service_name`](https://nginx.org/en/docs/ngx_otel_module.html#otel_service_name)
+- [`otel_span_attr`](https://nginx.org/en/docs/ngx_otel_module.html#otel_span_attr): set global span attributes that will be merged with the span attributes set in the [Observability extension](nginx-extensions.md#gateway-settings).
+
+In the future, this config will be extended to support other directives, such as those defined in the [NGINX Extensions Proposal](nginx-extensions.md#gateway-settings).
+
+## API, Customer Driven Interfaces, and User Experience
+
+The `GatewaySettings` API is a CRD that is a part of the `gateway.nginx.org` Group. It will be referenced in the `parametersRef` field of a GatewayClass. It will live at the cluster scope.
+
+For example, a `GatewaySettings` named `gw-settings` would be referenced as follows:
+
+```yaml
+kind: GatewayClass
+metadata:
+    name: nginx
+spec:
+    controllerName: gateway.nginx.org/nginx-gateway-controller
+    parametersRef:
+        group: gateway.nginx.org/v1alpha1
+        kind: GatewaySettings
+        name: gw-settings
+```
+
+Below is the Golang API for the `GatewaySettings` API:
+
+### Go
+
+```go
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type GatewaySettings struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the GatewaySettings.
+	Spec GatewaySettingsSpec `json:"spec"`
+
+	// Status defines the state of the GatewaySettings.
+	Status GatewaySettingsStatus `json:"status,omitempty"`
+}
+
+type GatewaySettingsSpec struct {
+	// OtelExporter specifies OpenTelemetry export parameters.
+	// +optional
+	OtelExporter *OtelExporter `json:"otelExporter,omitempty"`
+
+	// OtelServiceName is the "service.name" attribute of the Otel resource.
+	// +optional
+	OtelServiceName *string `json:"otelServiceName,omitempty"`
+
+	// OtelSpanAttributes are custom key/value attributes that are added to each span.
+	// Variables can be included in the values.
```

Comment:
Should we support anything that nginx supports?

Reply:
I'm not sure. We definitely can't support custom variables because there's no way for users to create custom variables right now. Supporting all nginx variables will be challenging from a validation perspective, plus I'm sure not all nginx variables are relevant here. Is there a subset that makes sense? Or can we just allow all `$` prefixed strings and let nginx do the validation for us? 

@pleshakov thoughts?

Reply:
> r can we just allow all $ prefixed strings and let nginx do the validation for us?

NGINX will fail to reload if referenced variable doesn't exist

Another option is to start with most common ones like here https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources/#actionproxyrequestheaderssetheader

It is not too difficult to support validation https://github.com/nginxinc/kubernetes-ingress/blob/0fc5b6e81433ee04bc8c66b2d6cc6c9a449b1e4f/pkg/apis/configuration/validation/virtualserver.go#L1104-L1168

When allowed to be configured by developers, not cluster operators, we will need to double check if there is a risk of exposing any NGINX data that is not meant to be exported by regular developers.

Reply:
Didn't we originally intend on supporting variables with one of our header filters, but then removed it?

Reply:
And how do we determine which variables are risky?

Do we provide a way in this CRD for cluster operators to deny certain directives from being specified, based on their discretion?

Reply:
> Didn't we originally intend on supporting variables with one of our header filters, but then removed it?

I think we decided to wait until support for variables is requested

> And how do we determine which variables are risky?

that'll be our judgement call. 
Or an option you mentioned.

> Do we provide a way in this CRD for cluster operators to deny certain directives from being specified, based on their discretion?

nope.

Reply:
Ok, so maybe for now we leave out variable support until requested, and we can revisit this at that point.